BalloonBox-Inc/near-oracle-contracts
# [NEAR-18] Smart contract - first release

**Summary**
This `dev` branch contains the first version of a smart contract for the NEARoracle DApp.

Ticket: [#18]

**Changes Proposed**
---
The contract does the following:
 - initializes an empty `LookupMap<K, V>` to store scores where
 K is the user's wallet address and V is a growable vector of scores 
 (scores are saved as structs: score #, timestamp, description)
 - stores a new user's score in a new growable `vec`
 - appends a new score for a returning user to their growable `vec`
 - [:heavy_dollar_sign: costs gas] queries the score history for any user --> returns an error if user doesn't exist
 - [gasless] query the state of the contract (contract owner, memory used, user counts, score counts,...) 
 - [gasless] checks whether a user owns a score record or not --> returns a boolean
 > note: data structures were chosen following the guideline in the `near_sdk` docs on [persistent collections](https://www.near-sdk.io/contract-structure/collections)

**QA Recommendations**
---
 1) Is the overall code structure correct (object declaration first, followed by trait implementations, ending with unit tests)?
 2) Are the main objects (structs and enums) understandable and logically named?
 3) Is there an initialization function for the main struct called `Contract`?
 4) Are there methods implemented for the main struct `Contract`?
 5) Are you able to store scores / query scores / query contract state?
 6) Are all major functionalities covered or did we leave out any functionality we should implement soon?
 7) Is the Rust logic sound and efficient? How can I improve my use of Rust logic?
 8) Can you seamlessly run the unit tests?
 9) Is the Rust contract reasonably lightweight? 
 As of now the contract itself takes up ~ 118000 bytes of memory (pretty good :weight_lifting_man: ! )
 and the memory cost for storing one score is ~ 239 bytes/score
